### PR TITLE
feat(40750): Refactoring to add inferred return type annotation to a function

### LINKED ISSUE #40750: Refactoring to add inferred return type annotation to a function
## Search Terms
- Refactoring, refactor
- type annotation
- quick fix, code action
- function, method
- return

## Suggestion
Add a refactoring that takes a function with no annotated return type, and inserts the inferred return type.

For example, starting with the code:

```ts
function foo() {
  return { a: 1, b: 2 };
}
```

If you trigger the refactoring on `foo`, it would become:

```ts
function foo(): { a: number; b: number; } {
  return { a: 1, b: 2 };
}
```

## Use Cases
A few example use cases:

- Give your public interfaces explicit typings.

- Updating a function from `() => T` to `() => T | undefined`. Today, you have to explicitly type out the return type and then add `| undefined`

- Extracting the return type to a new interface. Right now, this is a manual process. With this refactoring, this becomes a two step process: generate inferred return type and then extract to interface

- When migrating from JS to TS, many JS functions may not have explicit return types. Often you want to make the code stricter by having these return types

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #40750

### CODE REVIEW

File: src/services/refactors/inferFunctionReturnType.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        const signatures = type.getCallSignatures();
+        if (!signatures.length) return;
+
+        const returnType = typeChecker.typeToTypeNode(first(signatures).getReturnType(), declaration, NodeBuilderFlags.NoTruncation);
+        if (returnType) {
+            return { declaration, returnType };
+        }
+    }
+
+    function isConvertableDeclaration(node: Node): node is ConvertableDeclaration {
```
Comment (sandersn) [inline src/services/refactors/inferFunctionReturnType.ts:61]: is it convertable or convertible? help english sucks

Comment (sandersn) [reply src/services/refactors/inferFunctionReturnType.ts:61]: We have an error "Could not find convertible access expression", and the car is spelt 'convertible', so I'd go with that.

**Review by sandersn — APPROVED** [review]
Looks good. I think this is a useful refactoring. There is a more general, less useful refactor that adds a type annotation on any declaration that could have one, but doesn't need to. I don't think this PR should add that (1) because it's less useful (2) I can't remember how to make a single refactoring produce multiple fixes -- it may not be possible.

@jessetrinity can you take a look too? What is your opinion on the more general refactoring?

**Comment by sandersn** [comment]
@gabritto you might have opinions on the question of the more general refactoring.

File: src/services/refactors/inferFunctionReturnType.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/refactors/inferFunctionReturnType.ts:61]: @sandersn Thanks, I'll change.

**Comment by sandersn** [comment]
Sounds like nobody else wants to chime in, so I'll merge this after the 4.1 RC goes out.

**Review by jessetrinity — CHANGES_REQUESTED** [review]
Perhaps there should be some tests with multiple signatures:

```ts
// bad code - there should still be a test so that we at least know what the refactor does in this case, if doing nothing is too complicated.
function f(x: number) {
    return 1;
}
function f(x: number) {
    return "1";
}
```

```ts
// good code (?) - grabbing the first signature will add "number" as the return type, but we probably wanted "number | string"
function f(x: number): number;
function f(x: string): string;
function f(x: string | number) {
  return typeof(x) === "string" ? "a" : 1;
}
```

**Review by jessetrinity — APPROVED** [review]

**Pull request merged**